controller/api/public/http_server.go

```diff
@@ -0,0 +1,218 @@
+		return
+	}
+
+	rsp, err := h.grpcServer.Version(req.Context(), &protoRequest)
+	if err != nil {
+		writeErrorToHttpResponse(w, err)
+		return
+	}
+
+	err = writeProtoToHttpResponse(w, rsp)
+	if err != nil {
+		writeErrorToHttpResponse(w, err)
+		return
+	}
+}
+
+func (h *handler) handleSelfCheck(w http.ResponseWriter, req *http.Request) {
+	var protoRequest healthcheckPb.SelfCheckRequest
+	err := httpRequestToProto(req, &protoRequest)
+	if err != nil {
+		writeErrorToHttpResponse(w, err)
+		return
+	}
+
+	rsp, err := h.grpcServer.SelfCheck(req.Context(), &protoRequest)
+	if err != nil {
+		writeErrorToHttpResponse(w, err)
+		return
+	}
+
+	err = writeProtoToHttpResponse(w, rsp)
+	if err != nil {
+		writeErrorToHttpResponse(w, err)
+		return
+	}
+}
+
+func (h *handler) handleListPods(w http.ResponseWriter, req *http.Request) {
+	var protoRequest pb.Empty
+	err := httpRequestToProto(req, &protoRequest)
+	if err != nil {
+		writeErrorToHttpResponse(w, err)
+		return
+	}
+
+	rsp, err := h.grpcServer.ListPods(req.Context(), &protoRequest)
+	if err != nil {
+		writeErrorToHttpResponse(w, err)
+		return
+	}
+
+	err = writeProtoToHttpResponse(w, rsp)
+	if err != nil {
+		writeErrorToHttpResponse(w, err)
+		return
+	}
+}
+
+func (h *handler) handleTap(w http.ResponseWriter, req *http.Request) {
+	flushableWriter, err := newStreamingWriter(w)
+	if err != nil {
+		writeErrorToHttpResponse(w, err)
+		return
+	}
+
+	var protoRequest pb.TapRequest
+	err = httpRequestToProto(req, &protoRequest)
+	if err != nil {
+		writeErrorToHttpResponse(w, err)
+		return
+	}
+
+	server := tapServer{w: flushableWriter, req: req}
+	err = h.grpcServer.Tap(&protoRequest, server)
+	if err != nil {
+		writeErrorToHttpResponse(w, err)
+		return
+	}
+}
+
+type tapServer struct {
+	w   flushableResponseWriter
+	req *http.Request
+}
+
+func (s tapServer) Send(msg *common.TapEvent) error {
+	err := writeProtoToHttpResponse(s.w, msg)
+	if err != nil {
+		writeErrorToHttpResponse(s.w, err)
+		return err
+	}
+
+	s.w.Flush()
+	return nil
+}
+
+// satisfy the pb.Api_TapServer interface
+func (s tapServer) SetHeader(metadata.MD) error  { return nil }
+func (s tapServer) SendHeader(metadata.MD) error { return nil }
+func (s tapServer) SetTrailer(metadata.MD)       { return }
+func (s tapServer) Context() context.Context     { return s.req.Context() }
+func (s tapServer) SendMsg(interface{}) error    { return nil }
+func (s tapServer) RecvMsg(interface{}) error    { return nil }
+
+func fullUrlPathFor(method string) string {
+	return ApiRoot + ApiPrefix + method
+}
+
+func withTelemetry(baseHandler *handler) http.HandlerFunc {
+	counter := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "http_requests_total",
+			Help: "A counter for requests to the wrapped handler.",
+		},
+		[]string{"code", "method"},
+	)
+	prometheus.MustRegister(counter)
+	instrumentedHandler := promhttp.InstrumentHandlerCounter(counter, baseHandler)
+	return instrumentedHandler
```

Comment:
TIOLI No need for the `instrumentedHandler` local variable here -- you can just return the result of `InstrumentHandlerCounter` directly:

```
return promhttp.InstrumentHandlerCounter(counter, baseHandler)
```